### ISSUE TITLE
[Questions] How to provide custom error messages when a type is not available?

### ISSUE THREAD

**Issue by JosXa** [alecthomas/injector#167]
I would like to expose a `FooCredentials` dataclass to users as part of a library. 
Ignoring the fact that this could probably be done nicely by asking them to pass it to me via the `Module` initializer, how could I check whether a specific binding has been overriden by the user **at time of first access** (not at time of binding as user code might not have run yet)?

**Comment by jstasiak** [comment]
After reading the first sentence I was gonna say `binder.bind(FooCredentials, to=FooCredentials(loging='hello', password='world'))` but I can't understand the rest of your question at all. Can you provide some pseudocode of what you're trying to achieve?

**Comment by JosXa** [comment]
```python
class TickTickModule(Module):
    @provider
    def credentials(self, binder: Binder) -> TickTickCredentials:
        has_been_overridden_by_user = bool(binder.get_binding(TickTickCredentials))
        if not has_been_overridden_by_user:
            raise UnsatisfiedRequirement(
                "In order to work with this API, you should provide an instance of " +
                f"{TickTickCredentials.__name__} to injector!"
            )
        return binder.get_binding(TickTickCredentials)  # ok, they did their job
```

Here's a somewhat condensed version of [my eventual solution](https://gist.github.com/1b7479ca189adc74fe42f829248779af). Bear in mind that above is more of a toy example, as I __did__ after all go with the module initializer approach - but I've had some use cases where I wanted to check "ok has someone else been interfering with the library-provided dependencies here or are we clean?"

Do you you hear how I feel?

**Comment by JosXa** [comment]
Just wanted to note that WOW does it loosen up your architecture and drop you into that goldilock pit of success when you don't have to worry about all the wiring anymore.. Top-down containers are nice! (when compared to the global container approach of other libs)

**Comment by jstasiak** [comment]
Yes, sorry, it's clear now. There's no good way to get this now. I can't think of anything but an ugly hack like

```python
import injector


class TickTickCredentials:
    def __init__(self, username: str = '', password: str = '') -> None:
        if not username and not password:
            raise RuntimeError('In order to use this library you need to bind this')


injector.Injector().get(TickTickCredentials)
```

which gives you a custom exception if nobody constructs `TickTickCredentials` with non-empty `username` and `password`.
